java-checks/src/main/java/org/sonar/java/checks/EnumMapCheck.java

```diff
@@ -0,0 +1,114 @@
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.ParenthesizedTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeArguments;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+@Rule(
+  key = "S1640",
+  name = "Maps with keys that are enum values should be replaced with EnumMap",
+  tags = {"performance"},
+  priority = Priority.MAJOR)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.CPU_EFFICIENCY)
+@SqaleConstantRemediation("5min")
+public class EnumMapCheck extends BaseTreeVisitor implements JavaFileScanner {
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    if (tree.type().symbolType().isSubtypeOf("java.util.Map")) {
+      checkMap(tree);
+    } else {
+      super.visitVariable(tree);
+    }
+  }
+
+  private void checkMap(VariableTree tree) {
+    boolean returnTypeHasEnumKey = hasEnumKey(tree.type());
+    ExpressionTree initializer = tree.initializer();
+    if (initializer != null) {
+      initializer = removeParenthesis(initializer);
+      if (initializer.is(Tree.Kind.NEW_CLASS)) {
+        NewClassTree newClassTree = (NewClassTree) initializer;
+        if (newClassTree.symbolType().isSubtypeOf("java.util.HashMap") && (returnTypeHasEnumKey || hasEnumKey(newClassTree.identifier()))) {
+          addIssue(tree);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void visitNewClass(NewClassTree tree) {
+    if (tree.symbolType().isSubtypeOf("java.util.HashMap") && hasEnumKey(tree.identifier())) {
+      addIssue(tree);
+    } else {
+      super.visitNewClass(tree);
+    }
+  }
+
+  private static boolean hasEnumKey(TypeTree typeTree) {
+    if (typeTree.is(Tree.Kind.PARAMETERIZED_TYPE)) {
+      TypeArguments typeArguments = ((ParameterizedTypeTree) typeTree).typeArguments();
+      if (typeArguments.size() > 0) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use isEmpty() to check whether the collection is empty or not. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1155)